README.md
# node-oracledb version 4.0

The node-oracledb add-on for Node.js powers high performance Oracle
Database applications.

Use node-oracledb 4.0 to connect Node.js 8.16, 10.16, 12, or later, to Oracle Database.
Older versions of node-oracledb may work with older versions of Node.js.

Node-oracledb supports basic and advanced features of Oracle Database
and Oracle Client.  See the [homepage][4] for a list.

The node-oracledb module is open source and maintained by Oracle Corp.
It is stable, well documented, and has a comprehensive test suite.

## <a name="installation"></a> Installation

See [Getting Started with Node-oracledb][1] and [Quick Start Node-oracledb Installation][2].

## <a name="doc"></a> Documentation

See [Documentation for the Oracle Database Node.js Add-on][32] and the [CHANGELOG][33].

## <a name="examples"></a> Examples

See the [examples][30] directory.  Start with [examples/example.js][31].

## <a name="help"></a> Help

Issues and questions about node-oracledb can be posted on [GitHub][3] or
[Slack][5] ([link to join Slack][6]).

## <a name="testing"></a> Tests

To run the test suite see [test/README][34].

## <a name="contrib"></a> Contributing

Node-oracledb is an open source project. See [CONTRIBUTING][35] for
details.

Oracle gratefully acknowledges the contributions to node-oracledb that
have been made by the community.

## <a name="license"></a> License

Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved.

You may not use the identified files except in compliance with the Apache
License, Version 2.0 (the "License.")

You may obtain a copy of the License at
[http://www.apache.org/licenses/LICENSE-2.0][36].  Unless required by
applicable law or agreed to in writing, software distributed under the
License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied.

See the License for the specific language governing permissions and
limitations under the License.



[1]: https://oracle.github.io/node-oracledb/doc/api.html#getstarted
[2]: https://oracle.github.io/node-oracledb/INSTALL.html#quickstart
[3]: https://github.com/oracle/node-oracledb/issues
[4]: https://oracle.github.io/node-oracledb
[5]: https://node-oracledb.slack.com/
[6]: https://node-oracledb.slack.com/join/shared_invite/enQtNDU4Mjc2NzM5OTA2LTdkMzczODY3OGY3MGI0Yjk3NmQ4NDU4MTI2OGVjNTYzMjE5OGY5YzVkNDY4MWNkNjFiMDM2ZDMwOWRjNWVhNTg
[30]: https://github.com/oracle/node-oracledb/blob/master/examples
[31]: https://github.com/oracle/node-oracledb/blob/master/examples/example.js#L32
[32]: https://oracle.github.io/node-oracledb/doc/api.html
[33]: https://github.com/oracle/node-oracledb/blob/master/CHANGELOG.md
[34]: https://github.com/oracle/node-oracledb/blob/master/test/README.md
[35]: https://github.com/oracle/node-oracledb/blob/master/CONTRIBUTING.md
[36]: http://www.apache.org/licenses/LICENSE-2.0


CHANGELOG.md
# Change Log

## node-oracledb v4.0.1 (19 Aug 2019)

- Fixed a regression causing a segfault when setting `oracledb.connectionClass`
  and not creating a pool ([ODPI-C change](https://github.com/oracle/odpi/commit/f945355f3e58e7337dd798cba0404ab5755f0692)).

- Fixed a regression when enumerable properties were added to
  `Object.prototype`.
  ([#1129](https://github.com/oracle/node-oracledb/issues/1129)).

- Fixed a regression with missing `metaData` from `connection.getStatementInfo()`

- Fixed crashes with spurious subscription (e.g. CQN) notifications, and when
  unsubscribing an invalid subscription.

- A more meaningful error is returned when calling `connection.subscribe()` with
  SQL that is not a SELECT statement ([ODPI-C
  change](https://github.com/oracle/odpi/commit/f95846bef6cf70e8114cbbb59ca04fbe2e7a3903))

- Fixed passing DbObjects and JavaScript objects as the `payload` attribute for
  AQ message enqueues when using an object queue.

- Made the error message for AQ `queue.deqMany(0)` the same NJS-005 given when a
  negative number is used.

- Fixed a compilation warning seen on Windows.

- Improve portability of buildbinary.js, a package creation script ([#1129](https://github.com/oracle/node-oracledb/issues/1129)).

## node-oracledb v4.0.0 (25 Jul 2019)

- Refactored the node-oracledb implementation to use
  [N-API](https://nodejs.org/api/n-api.html) in place of
  [NAN](https://github.com/nodejs/nan).
    - Node-oracledb 4 requires Node.js 8.16 or Node.js 10.16, or higher.  Node.js 8.16, 10.16, 11.12 and 12 contain an important N-API performance fix.
    - N-API allows node-oracledb binaries to be portable between Node.js versions on a given operating system, subject to N-API compatibility.  Node-oracledb uses N-API version 4.
    - Oracle Client libraries are still required at runtime.  These can be from Oracle Instant Client, the full Oracle Client, or an Oracle Database installation.
    - The string representation of classes has changed to `[object Object]` as a consequence of using N-API.  Use `Object.getPrototypeOf()` to get class information.
    - The C compiler required for building from source code no longer needs C++11 compatibility.  The node-oracledb source code is now pure C.

- Added support for querying and binding [Oracle Database Objects and
  Collections](https://oracle.github.io/node-oracledb/doc/api.html#objects).

- Added support for [Oracle Advanced Queuing (AQ)](https://oracle.github.io/node-oracledb/doc/api.html#aq):

    - Added support for "RAW" queues, allowing String and Buffer
      messages to be used.

    - Added support for object queues, allowing Oracle Database object
      messages to be used.

    - Added support for notifications with `oracledb.SUBSCR_NAMESPACE_AQ`.

- Added support for [Implicit
  Results](https://oracle.github.io/node-oracledb/doc/api.html#implicitresults),
  allowing query results to be returned from PL/SQL without needing
  parameters or bind variables.

- Added asynchronous method
  [`lob.getData()`](https://oracle.github.io/node-oracledb/doc/api.html#lobgetdata)
  to return all data from a Lob readable stream.

- Added a new `dbTypeName` attribute to [`extendedMetaData`
  output](https://oracle.github.io/node-oracledb/doc/api.html#execmetadata).
  It contains the name of the type the column has in the database,
  such as "VARCHAR2".

- Enhanced BIND_IN of PL/SQL Collection Associative Arrays (Index-by)
  so a bind definition object can be omitted (see
  [#1039](https://github.com/oracle/node-oracledb/issues/1039)).

- Continuous Query Notification (CQN) Improvements:

    - Added support for getting the [registration
      id](https://oracle.github.io/node-oracledb/doc/api.html#consubscribecallback)
      for CQN subscriptions.

    - Added support and message type constants for database startup
      and shutdown events.

    - Fixed a crash that occurred when unsubscribing from CQN while
      notifications were ongoing ([ODPI-C
      change](https://github.com/oracle/odpi/commit/b96b11b7fe58f32f011c7f7419555e40268d5bf4)).

- Added
  [`connection.currentSchema`](https://oracle.github.io/node-oracledb/doc/api.html#propconncurrentschema)
  for setting the schema qualifier to be used when a qualifier is
  omitted in SQL statements.  This is an efficient alternative to
  `ALTER SESSION SET CURRENT_SCHEMA`.

- Renumbered [node-oracledb Type
  Constants](https://oracle.github.io/node-oracledb/doc/api.html#oracledbconstantsnodbtype)
  and [Oracle Database Type
  Constants](https://oracle.github.io/node-oracledb/doc/api.html#oracledbconstantsdbtype)
  to allow for future enhancements.

- Introduced [Query `outFormat`
  Constants](https://oracle.github.io/node-oracledb/doc/api.html#oracledbconstantsoutformat)
  `oracledb.OUT_FORMAT_ARRAY` and `oracledb.OUT_FORMAT_OBJECT`.  The
  previous constants `oracledb.ARRAY` and `oracledb.OBJECT` are
  deprecated but still usable.

- Improved the performance of [`oracledb.outFormat`](#propdboutformat) mode `oracledb.OUT_FORMAT_OBJECT`.

- Improved the fetch performance of LOBs in some cases by reducing the
  number of round-trips required between node-oracledb and Oracle
  Database ([ODPI-C
  change](https://github.com/oracle/odpi/commit/58e6a07ff5bb428a09068456ef5231884fcb77db)).

- Change the
  [`events`](https://oracle.github.io/node-oracledb/doc/api.html#propdbevents)
  default to *true*.

- Updated the JavaScript syntax in class implementations.

- Class methods are now configurable.  For example via
  `Object.defineProperty`.

- Error handling changes:

    - Corrected the error message returned when invalid types are used for
      boolean options.

    - Standardized error messages for incorrect function parameters.  Now
      NJS-005 and NJS-007 are used in place of NJS-006 and NJS-008,
      respectively.


lib/oracledb.js
// Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved

//-----------------------------------------------------------------------------
//
// You may not use the identified files except in compliance with the Apache
// License, Version 2.0 (the "License.")
//
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-----------------------------------------------------------------------------

'use strict';

const nodbUtil = require('./util.js');

// This version of node-oracledb works with Node.js 8.16, 10.16 or
// later.  The test stops hard-to-interpret runtime errors and crashes
// with older Node.js versions.  Also Node.js 8.16 and 10.16 (and
// 12.0) contain an important N-API performance regression fix.  If
// you're using the obsolete Node.js 9 or 11 versions, you're on your
// own regarding performance and functionality
let vs = process.version.substring(1).split(".").map(Number);
if (vs[0] < 8 || (vs[0] === 8 && vs[1] < 16)) {
  throw new Error(nodbUtil.getErrorMessage('NJS-069', nodbUtil.PACKAGE_JSON_VERSION, "8.16"));
} else if ((vs[0] === 10 && vs[1] < 16)) {
  throw new Error(nodbUtil.getErrorMessage('NJS-069', nodbUtil.PACKAGE_JSON_VERSION, "10.16"));
}

const AqDeqOptions = require('./aqDeqOptions.js');
const AqEnqOptions = require('./aqEnqOptions.js');
const AqMessage = require('./aqMessage.js');
const AqQueue = require('./aqQueue.js');
const BaseDbObject = require('./dbObject.js');
const Connection = require('./connection.js');
const Lob = require('./lob.js');
const Pool = require('./pool.js');
const ResultSet = require('./resultset.js');
const SodaDatabase = require('./sodaDatabase.js');
const SodaCollection = require('./sodaCollection.js');
const SodaDocCursor = require('./sodaDocCursor.js');
const SodaDocument = require('./sodaDocument.js');
const SodaOperation = require('./sodaOperation.js');

let poolCache = {};
let tempUsedPoolAliases = {};
const defaultPoolAlias = 'default';

//  Load the Oracledb binary

const binaryLocations = [
  '../' + nodbUtil.RELEASE_DIR + '/' + nodbUtil.BINARY_FILE,  // pre-built binary
  '../' + nodbUtil.RELEASE_DIR + '/' + 'oracledb.node',       // binary built from source
  '../build/Debug/oracledb.node'                              // debug binary
];

let oracledbCLib;
for (let i = 0; i < binaryLocations.length; i++) {
  try {
    oracledbCLib = require(binaryLocations[i]);
    break;
  } catch(err) {
    if (err.code !== 'MODULE_NOT_FOUND' || i == binaryLocations.length - 1) {
      let nodeInfo;
      if (err.code === 'MODULE_NOT_FOUND') {
        // none of the three binaries could be found
        nodeInfo = `\n  Looked for ${binaryLocations.map(x => require('path').resolve(__dirname, x)).join(', ')}\n  ${nodbUtil.getInstallURL()}\n`;
      } else {
        nodeInfo = `\n  Node.js require('oracledb') error was:\n  ${err.message}\n  ${nodbUtil.getInstallHelp()}\n`;
      }
      throw new Error(nodbUtil.getErrorMessage('NJS-045', nodeInfo));
    }
  }
}


class OracleDb {

  constructor() {
    this.queueTimeout = 60000;
    this.Promise = global.Promise;
  }

  // extend class with promisified functions
  _extend(oracleDbInst) {
    this.getConnection = nodbUtil.promisify(oracleDbInst, getConnection);
    this.createPool = nodbUtil.promisify(oracleDbInst, createPool);
  }

  // retrieves a pool from the pool cache (synchronous method)
  getPool(poolAlias) {
    let pool;

    nodbUtil.assert(arguments.length < 2, 'NJS-009');

    if (poolAlias) {
      nodbUtil.assert(typeof poolAlias === 'string' || typeof poolAlias === 'number', 'NJS-005', 1);
    }

    poolAlias = poolAlias || defaultPoolAlias;

    pool = poolCache[poolAlias];

    if (!pool) {
      throw new Error(nodbUtil.getErrorMessage('NJS-047', poolAlias));
    }

    return pool;
  }

}

// Oracledb functions and classes

// oracledbCLib.OracleDb.prototype.newLob = function(iLob) {
//   return new Lob(iLob, null, oracledbInst);
// };

// This createPool function is used the override the createPool method of the
// Oracledb class, which is defined in the C layer. The override allows us to do
// things like extend out the pool instance prior to passing it to the caller.
function createPool(poolAttrs, createPoolCb) {
  const self = this;
  let sessionCallback;
  let poolAlias;

  // Initial argument count and type checks are done first and throw in the same
  // call stack.
  nodbUtil.assert(arguments.length === 2, 'NJS-009');
  nodbUtil.assert(nodbUtil.isObject(poolAttrs), 'NJS-005', 1);
  nodbUtil.assert(typeof createPoolCb === 'function', 'NJS-005', 2);

  // Additional validations should pass errors via the callback. Need to ensure
  // that errors are raised prior to actually creating the pool via _createPool.
  if (poolAttrs.poolAlias !== undefined) {
    if (typeof poolAttrs.poolAlias !== 'string' || poolAttrs.poolAlias.length === 0) {
      createPoolCb(new Error(nodbUtil.getErrorMessage('NJS-004', 'poolAttrs.poolAlias')));
      return;
    }

    poolAlias = poolAttrs.poolAlias;
  } else if (poolAttrs.poolAlias === undefined
    && !poolCache[defaultPoolAlias]
    && !tempUsedPoolAliases[defaultPoolAlias]
  ) {
    poolAlias = defaultPoolAlias;
  }

  if (poolCache[poolAlias] || tempUsedPoolAliases[poolAlias]) {
    createPoolCb(new Error(nodbUtil.getErrorMessage('NJS-046', poolAlias)));
    return;
  }

  // Retain local callback for fixing up tags on connections acquired from the
  // pool, if applicable; this value can either be a function which will be
  // called when a connection is acquired from the pool which doesn't have the
  // requested tag or will be a string defining a PL/SQL procedure which will
  // be called by the database when a connection acquired from the pool doesn't
  // have the requested tag
  sessionCallback = poolAttrs.sessionCallback;
  if (typeof poolAttrs.sessionCallback === 'function') {
    poolAttrs = Object.assign({}, poolAttrs);
    delete poolAttrs.sessionCallback;
  }

  // Need to prevent another call in the same stack from succeeding, otherwise
  // two pools could be created with the same poolAlias and the second one that
  // comes back would overwrite the first in the cache.
  if (poolAlias) {
    tempUsedPoolAliases[poolAlias] = true;
  }

  self._createPool(poolAttrs, function(err, poolInst) {
    if (err) {
      // We need to free this up since the creation of the pool failed.
      if (poolAlias) {
        delete tempUsedPoolAliases[poolAlias];
      }

      if (err.message.match(/DPI-1047/)) {
        err.message += "\n" + nodbUtil.getInstallHelp();
      }
      createPoolCb(err);

      return;
    }

    if (poolAlias) {
      poolCache[poolAlias] = poolInst;

      // It's now safe to remove this alias from the tempUsedPoolAliases.
      delete tempUsedPoolAliases[poolAlias];
    }

    poolAttrs.sessionCallback = sessionCallback;
    poolInst._setup(poolAttrs, poolAlias, self);

    poolInst.on('_after_close', function() {
      const pool = this;

      if (pool.poolAlias) {
        delete poolCache[pool.poolAlias];
      }
    });

    createPoolCb(null, poolInst);
  });
}

// This getConnection function is used the override the getConnection method of the
// Oracledb class, which is defined in the C layer. The override allows us to do
// things like extend out the connection instance prior to passing it to the caller.
function getConnection(a1, a2) {
  const self = this;
  let pool;
  let poolAlias;
  let connAttrs = {};
  let getConnectionCb;

  nodbUtil.assert(arguments.length < 3, 'NJS-009');

  // Verify the number and types of arguments, then initialize the local poolAlias,
  // connAttrs, and getConnectionCb variables based on the arguments.
  switch (arguments.length) {
    case 1:
      nodbUtil.assert(typeof a1 === 'function', 'NJS-005', 1);

      poolAlias = defaultPoolAlias;
      getConnectionCb = a1;

      break;
    case 2:
      nodbUtil.assert(typeof a1 === 'string' || nodbUtil.isObject(a1), 'NJS-005', 1);
      nodbUtil.assert(typeof a2 === 'function', 'NJS-005', 2);

      if (typeof a1 === 'string') {
        poolAlias = a1;
      } else if (nodbUtil.isObject(a1)) {
        connAttrs = a1;

        if (connAttrs.poolAlias) {
          poolAlias = connAttrs.poolAlias;
        }
      }

      getConnectionCb = a2;

      break;
  }

  // Proceed to execution based on values in local variables. Look for the poolAlias
  // first and only attempt to use connAttrs if the poolAlias isn't set.
  if (poolAlias) {
    pool = poolCache[poolAlias];

    if (!pool) {
      getConnectionCb(new Error(nodbUtil.getErrorMessage('NJS-047', poolAlias)));
      return;
    }

    pool.getConnection(connAttrs, getConnectionCb);
  } else {
    self._getConnection(connAttrs, function(err, connInst) {
      if (err) {
        if (err.message.match(/DPI-1047/)) {
          err.message += "\n" + nodbUtil.getInstallHelp();
        }
        getConnectionCb(err);
        return;
      }

      getConnectionCb(null, connInst);
    });
  }
}

// create instance which will be exported
let oracleDbInst = new OracleDb();

// add classes to prototype
let proto = Object.getPrototypeOf(oracleDbInst);
proto.OracleDb = OracleDb;
proto.AqDeqOptions = AqDeqOptions;
proto.AqEnqOptions = AqEnqOptions;
proto.AqMessage = AqMessage;
proto.AqQueue = AqQueue;
proto.BaseDbObject = BaseDbObject;
proto.Connection = Connection;
proto.Lob = Lob;
proto.Pool = Pool;
proto.ResultSet = ResultSet;
proto.SodaDatabase = SodaDatabase;
proto.SodaCollection = SodaCollection;
proto.SodaDocCursor = SodaDocCursor;
proto.SodaDocument = SodaDocument;
proto.SodaOperation = SodaOperation;

// call C to extend classes
oracledbCLib.init(oracleDbInst);

module.exports = oracleDbInst;
